lib/inc/internal/util/aix/odm.hpp

```diff
@@ -0,0 +1,263 @@
+#include <odmi.h>
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+namespace facter { namespace util { namespace aix {
+
+    // This is a singleton class that handles initialization and
+    // termination of the ODM library. As long as any other class holds a
+    // reference to this one, ODM will remain initialized.
+    //
+    // This class also provides static helpers for interacting with the ODM
+    class odm {
+    public:
+         typedef std::shared_ptr<odm> ptr;
+         static ptr get() {
+             static std::weak_ptr<odm> self;
+
+             auto result = self.lock();
+             if(!result) {
+                 result = ptr(new odm);
+                 self = result;
+             }
+             return result;
+         }
+
+         odm(const odm&) = delete;
+         odm& operator=(const odm&) = delete;
+
+         ~odm() {
+             odm_terminate();
+         }
+
+        // unlike just about every other error code to string function in the
+        // world, odm_err_msg doesn't just return its pointer - it wants a
+        // char** argument to put it in. This wraps that ugly API in a more
+        // strerror()-like interface for ease of use.
+         static const char* error_string() {
+             static char* msg;
```

Comment:
So repeated calls to error_string will overwrite the value. I suppose that's fine, if you want to keep a copy put it in a std::string.